Implemented mock data node

alberto-escobar opened the pull request:
<!-- Remember to add the relevant reviewers, assignees, and labels, and create this PR as a draft if it is a work in progress. -->

### Description
<!-- Replace _issue_ with the issue number that this PR resolves, or delete the line and add this PR to the Software project if there is no related issue. -->
- Resolves #451 
<!-- Describe what was done in this PR if there is no related issue, or the related issue's description is not sufficient. -->


### Verification
<!-- List the steps that were taken to verify that the changes introduced by this PR function as desired and without side effects. -->
- Mock data appearing on console when running package and previous warning messages regarding not all subscribers are active in the physics engine node have diminished.
- tested node with and without mock data node cli argument `enable-mock-data:=true` to ensure argument logic is correct

### Resources
<!-- Link to any resources that are relevant to this PR. -->
- my big ol'brain

SPDonaghy commented:
Resources:
![image](https://github.com/user-attachments/assets/ed86ef60-0cec-4699-b6d6-40fed122d453)

alberto-escobar commented:
> Resources: ![image](https://private-user-images.githubusercontent.com/118148642/382511215-ed86ef60-0cec-4699-b6d6-40fed122d453.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzA1ODk2NjMsIm5iZiI6MTczMDU4OTM2MywicGF0aCI6Ii8xMTgxNDg2NDIvMzgyNTExMjE1LWVkODZlZjYwLTBjZWMtNDY5OS1iNmQ2LTQwZmVkMTIyZDQ1My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMTAyJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTEwMlQyMzE2MDNaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wODljZTIwZjU4ZjFkNDdjMmU4Y2M5Y2FmYTYwYzNkYTljMTZmODFiODhiYWM5NDY5ZmVhNTgyMjg2ZGRlYWU4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.sf5xkoVIxx3eThAK8ojCOQT2PnIeTijK2P6l1H9RRig)

HAHAHAHAAHAHA

alberto-escobar commented:
new CLI arguement for boat_simulator:
`enable-mock-data:=true`

eomielan commented on src/boat_simulator/boat_simulator/common/constants.py:
```diff
@@ -91,6 +91,9 @@ class BoatProperties:
 # CLI argument name for data collection option
 DATA_COLLECTION_CLI_ARG_NAME = "--enable-data-collection"
 
+# CLI argument name for data collection option
```
Please update comment here

eomielan commented on src/boat_simulator/boat_simulator/common/constants.py:
```diff
@@ -91,6 +91,9 @@ class BoatProperties:
 # CLI argument name for data collection option
 DATA_COLLECTION_CLI_ARG_NAME = "--enable-data-collection"
 
+# CLI argument name for data collection option
+MOCK_DATA_CLI_ARG_NAME = "--enable-mock-data"
+
```
Please update comment here

alberto-escobar replied:
done

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,91 @@
+
+
+def main(args=None):
+    rclpy.init(args=args)
+    node = MockDataNode()
+    if is_mock_data_enabled():
+        try:
+            rclpy.spin(node)
+        finally:
+            rclpy.shutdown()
```
Small optimization here, you can likely avoid creating an instance of `MockDataNode` until after checking if it is enabled. That is, move `node = MockDataNode()` inside the `try` statement.

alberto-escobar replied:
changed it so the instantiation is within the try block.

alberto-escobar commented:
> Nice work so far! It would be great if you could also configure the mock data node using the `globals.yml` file. Ideally, we should be able to specify which topics to publish mock data to, the exact data to publish, and the frequency at which it’s published. For an example of how this configuration was added for the data collection node, please refer to #222. Within the node, you’ll need to use [ROS parameters](https://docs.ros.org/en/humble/Tutorials/Beginner-Client-Libraries/Using-Parameters-In-A-Class-Python.html) to access the values defined in `globals.yml`. You can see how these parameters are used in the data collection node in this section:
> 
> https://github.com/UBCSailbot/sailbot_workspace/blob/6c5d7f3f2f51d0d3ac93baf2f5cb80aa5af30302/src/boat_simulator/boat_simulator/nodes/data_collection/data_collection_node.py#L80

I have added the following parameters to `globals.yaml`:
```
mock_data_node:
   ros__parameters:
      qos_depth: 10
      mock_desired_heading: True
      mock_sail_trim_tab: False
      pub_period_sec: 1.0
```
and added logic so the node only publishes mock data to the topics that have a true value in `globals.yaml`

eomielan commented on src/global_launch/config/globals.yaml:
```diff
@@ -76,3 +76,10 @@ data_collection_node:
       bag: false
       json: true
       write_period_sec: 0.5
+
+mock_data_node:
+   ros__parameters:
+      qos_depth: 10
+      mock_desired_heading: True
+      mock_sail_trim_tab: True
+      pub_period_sec: 5.0
```
Please update the `README` in the same directory with these new paramaters

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,131 @@
+
+import boat_simulator.common.constants as Constants
+
+# based on following:
+# https://docs.ros.org
+# /en/humble/Tutorials/Beginner-Client-Libraries/Writing-A-Simple-Py-Publisher-And-Subscriber.html
+# the purpose of this node is to publish to all topics that there is
+# subscribers in physics engine node so send goal code can work.
```
Could you move this to a docstring for the `MockDataNode` class

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,131 @@
+        msg.heading = helper_heading
+
+        self.desired_heading_pub.publish(msg)
+        self.get_logger().info(
+            f"Publishing to {self.desired_heading_pub.topic} "
+            + f"a mock desired heading of {heading} degrees"
+        )
```
No need for `+` here. https://stackoverflow.com/a/48881390

alberto-escobar replied:
huh, you learn something new everyday.

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,131 @@
+        msg.trim_tab_angle_degrees = trim_tab_angle_degrees
+
+        self.sail_trim_tab_angle_pub.publish(msg)
+        self.get_logger().info(
+            f"Publishing to {self.sail_trim_tab_angle_pub.topic} "
+            + f"a mock trim tab angle of {trim_tab_angle_degrees} degrees"
+        )
```
Same here

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,131 @@
+
+    def publish_mock_desired_heading(self):
+        """Publishes mock wind sensor data."""
+        heading = random.uniform(-179.99, 180.0)
```
Could you please make the interval configurable through `globals.yaml` as well. This would make it easier to adjust on the fly and help avoid the need to rebuild the boat simulator package every time it's changed.

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,131 @@
+
+    def publish_mock_sail_trim_tab_angle(self):
+        """Publishes mock wind sensor data."""
+        trim_tab_angle_degrees = random.uniform(-40, 40)
```
Same for the interval here

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,145 @@
+        if self.mock_sail_trim_tab:
+            self.publish_mock_sail_trim_tab_angle()
+
+    def publish_mock_desired_heading(self):
+        """Publishes mock wind sensor data."""
```
Please update this docstring to mock desired heading.

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,145 @@
+            f'a mock desired heading of {heading} degrees'
+        )
+
+    def publish_mock_sail_trim_tab_angle(self):
+        """Publishes mock wind sensor data."""
```
Please also update this docstring.

eomielan commented on src/global_launch/config/README.md:
```diff
@@ -250,7 +250,7 @@ specified within an array: one for the `x` component, and one for the `y` compon
 - _Datatype_: `double` array, length 2
 - _Range_: `(0.0, MAX_DOUBLE)`
-    - If a standard deviation of zero is desired, then consider using the constant generator instead.
+- If a standard deviation of zero is desired, then consider using the constant generator instead.
```
Could you please add the indentation back.

eomielan commented on src/global_launch/config/globals.yaml:
```diff
@@ -76,3 +76,13 @@ data_collection_node:
+      mock_desired_heading_upper_bound: 180.00
+      mock_sail_trim_tab: True
+      mock_sail_trim_tab_lower_bound: -40.0
+      mock_sail_trim_tab_desired_heading_upper_bound: 40
```
Should this be called `mock_sail_trim_tab_upper_bound`?

eomielan commented on src/boat_simulator/boat_simulator/nodes/mock_data/mock_data_node.py:
```diff
@@ -0,0 +1,145 @@
+
+
+class MockDataNode(Node):
+    """the purpose of this node is to publish to all topics that there is subscribers in physics.
+    engine node so send goal code can work.
```
Could you please remove early `.` on first line here

alberto-escobar commented:
tested node and it is working.

alberto-escobar commented:
tested node and it works

eomielan reviewed (approved):
LGTM nice work!

alberto-escobar commented:
@SPDonaghy @jamenkaye @FireBoyAJ24 can y'all approve this PR? y'all got added as a reviewer for some reason but Eric has approved this PR today.
